Sonatype release API auth now requires API token, rejects Nexus UI username & password

As of mid-June 2024, the Sonatype release API now [_requires_](https://central.sonatype.org/publish/generate-token/) token authentication, and the old authentication method using the Nexus UI username & password no longer works. Sonatype credentials now look like this:

* **Nexus UI username & password**: now only used for accessing the Nexus console UI at https://oss.sonatype.org/ - username & password are both chosen by the user. Username is fixed (ie **non-revokable**), and can be public
* **Token**: supplied in two parts (effectively passed as 'username' & 'password' in the API) as two randomised strings (the user can't choose these! 👍 ). Can be rotated ('reset') without changing the Nexus UI password. As it is now a meaningful secret (ie one that **can be revoked**), the **username should be handled as a secret**.

Sonatype communicated by email:

> To configure a publisher’s plugin authentication you would need to update your plugin settings to use [a user token](https://central.sonatype.org/publish/generate-token/) instead of the Nexus UI username and password login.

<img width="1456" alt="image" src="https://github.com/guardian/gha-scala-library-release-workflow/assets/52038/66d7f60f-b8ab-4315-96b2-5cb3b669ac6f">

This breaks library release for us! [Attempting to release](https://github.com/guardian/mobile-apps-api-models/actions/runs/9584722337/job/26429213001#step:5:32) with a Nexus UI username & password login will get this `401 Content access is protected by token` error:

```
Caused by: java.io.IOException: Unexpected server response: 401 Content access is protected by token
	at org.sonatype.spice.zapper.client.hc4.Hc4Client.upload(Hc4Client.java:107)
	at org.sonatype.spice.zapper.client.hc4.Hc4Track.call(Hc4Track.java:38)
	at org.sonatype.spice.zapper.client.hc4.Hc4Track.call(Hc4Track.java:23)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)  - (Sonatype.scala:443)
```


This follows on from Sonatype's announcement in [January 2024](https://central.sonatype.org/news/20240109_issues_sonatype_org_deprecation/#support-requests), where they began to actively discouraging the legacy username & password method of authentication.

In the new token-based new scheme, the token is still split into a username/password format, and both are randomised strings, making the username portion a meaningful secret (ie one that can be revoked) and so worthy of being treated as a secret - on the same level of sensitivity as the password.

## See also

* https://chat.google.com/room/AAAAag0I08g/Oi-XzrQ-GWM/Oi-XzrQ-GWM

--- a/.github/workflows/reusable-release.yml
+++ b/.github/workflows/reusable-release.yml
@@ -20,14 +20,9 @@ on:
         default: 'oss.sonatype.org' # The default host is going to be whatever "com.gu" is using
         required: false # ...but if you're not the Guardian, you'll want to set this explicitly
         type: string
-      SONATYPE_USERNAME:
-        description: 'Sonatype username'
-        default: 'guardian.automated.maven.release' # Only for use by the Guardian!
-        required: false # Must be supplied if used by a non-Guardian project
-        type: string
     secrets:
-      SONATYPE_PASSWORD:
-        description: 'Password for the SONATYPE_USERNAME account - used to authenticate when uploading artifacts'
+      SONATYPE_TOKEN:
+        description: 'Sonatype authentication token, colon-separated (username:password) - https://central.sonatype.org/publish/generate-token/'
         required: true
       PGP_PRIVATE_KEY:
         description:
@@ -416,9 +411,10 @@ jobs:
           cache: sbt # the issue described in https://github.com/actions/setup-java/pull/564 doesn't affect this step (no version.sbt)
       - name: Release
         env:
-          SONATYPE_USERNAME: ${{ inputs.SONATYPE_USERNAME }}
-          SONATYPE_PASSWORD: ${{ secrets.SONATYPE_PASSWORD }}
+          SONATYPE_TOKEN: ${{ secrets.SONATYPE_TOKEN }}
         run: |
+          export SONATYPE_USERNAME="${SONATYPE_TOKEN%%:*}" # See https://github.com/xerial/sbt-sonatype/pull/62
+          export SONATYPE_PASSWORD="${SONATYPE_TOKEN#*:}"
           sbt "sonatypeBundleRelease"
 
   github-release:

--- a/docs/credentials/generating-credentials.md
+++ b/docs/credentials/generating-credentials.md
@@ -3,9 +3,25 @@
 Normally you'll be using [shared organisation-wide credentials](supplying-credentials.md),
 but if you need to rotate those credentials, or just create some new ones for your organisation:
 
-## Updating a Sonatype OSSRH user's password
+## Updating a Sonatype OSSRH Token username & password
 
-See [Sonatype's instructions](https://central.sonatype.org/faq/ossrh-password/).
+As of [January 2024](https://central.sonatype.org/news/20240109_issues_sonatype_org_deprecation/#support-requests),
+Sonatype is actively discouraging the legacy username & password method of authentication, recommending
+[token authentication](https://central.sonatype.org/publish/generate-token/)
+(see link for token-regenerating instructions).
+
+Note these points:
+
+* The token is in a colon:separated username/password format, and _both_ username & password are randomised & revocable
+  secret strings.
+* Tokens generated on either https://oss.sonatype.org/ or https://s01.oss.sonatype.org/ will be _different_, and
+  **a token generated on one will not work on the other**. So, eg, if your `SONATYPE_CREDENTIAL_HOST` is `s01.oss.sonatype.org`,
+  you'll need to use a token _generated_ on `s01.oss.sonatype.org`. Remember that the `SONATYPE_CREDENTIAL_HOST` you
+  use is [dictated](https://github.com/xerial/sbt-sonatype/pull/461) by which Sonatype OSSRH server your **profile**
+  is hosted on.  
+  **Guardian developers:** currently the Guardian's `com.gu` profile is hosted on `oss.sonatype.org`, so the token we
+  use must be generated [there](https://oss.sonatype.org/), logged in with the `guardian.automated.maven.release`
+  account.
 
 ## Generating a new PGP key
 

--- a/docs/credentials/supplying-credentials.md
+++ b/docs/credentials/supplying-credentials.md
@@ -20,7 +20,7 @@ has _access_ to those secrets.
 to grant repos access to the necessary Organisation secrets - you need to raise a PR (like [this example PR](https://github.com/guardian/github-secret-access/pull/24))
 which will grant access to these:
 
-* `AUTOMATED_MAVEN_RELEASE_SONATYPE_PASSWORD`
+* `AUTOMATED_MAVEN_RELEASE_SONATYPE_TOKEN`
 * `AUTOMATED_MAVEN_RELEASE_PGP_SECRET`
 * `AUTOMATED_MAVEN_RELEASE_GITHUB_APP_PRIVATE_KEY`
 

